### PR TITLE
Add zippering support

### LINKED ISSUE #61418: Open source builds do not support MacCatalyst target
The open source toolchain distributions do not have the appropriate libraries to support the MacCatalyst target. 

For example, trying to build a catalyst app using the 5.7 release toolchain downloaded from swift.org results in:

```
ld64.lld: error: ../../third_party/swift-toolchain/usr/lib/swift/macosx/libswiftCompatibilityConcurrency.a(CompatibilityConcurrency.cpp.o) has platform macOS, which is different from target platform macCatalyst
```

The above is an excerpt from Chromium's catalyst buildbot: https://logs.chromium.org/logs/chromium/buildbucket/cr-buildbucket/8801593085733033281/+/u/compile/raw_io.output_text_failure_summary_

**Environment (please fill out the following information)**
 - OS:  all
 - Xcode Version/Tag/Branch: Xcode14

### PR DESCRIPTION
**Pull request by etcwilde**
Setting the `SWIFT_ENABLE_MACCATALYST` flag will build the compatibility libraries with zippering enabled.

Note that AppleClang and LLVM Clang use different flags to write zippered files. The Swift stdlib build doesn't know what clang we're using, so we can't ask it for the compiler ID to determine which whether we're using AppleClang or LLVM clang, hence the nasty `execute_process` trick to figure out whether the compiler that's actually compiling the stdlib knows about the flag or not.

This should fix #61418.

### CODE REVIEW

**Comment by etcwilde** [comment]
@compnerd I'm pretty sure that this shouldn't break Windows since the runtimes should only be built with the special clang, which will know about the catalyst bundling and because we can't build for catalyst on Windows anyway. But I'm not 100% certain. I'd appreciate your input.

**Comment by etcwilde** [comment]
@swift-ci please test

File: stdlib/cmake/modules/AddSwiftStdlib.cmake [hunk 1]
```diff
@@ -82,7 +82,28 @@ function(_add_target_variant_c_compile_link_flags)
       DEPLOYMENT_VERSION "${DEPLOYMENT_VERSION}")
     list(APPEND result "-target" "${target}")
     if(target_variant)
-      list(APPEND result "-target-variant" "${target_variant}")
+      # Check if the C compiler supports `-target-variant` flag
+      # TODO (etcwilde): This is a massive hack to deal with the fact that we
+      # are lying to cmake about what compiler is being used. Normally we could
+      # use `check_cxx_compiler_flag` here. Unfortunately, that uses a different
```
Comment (compnerd) [inline stdlib/cmake/modules/AddSwiftStdlib.cmake:88]:
```suggestion
      # use `check_c_compiler_flag` here. Unfortunately, that uses a different
```

File: stdlib/cmake/modules/AddSwiftStdlib.cmake [hunk 2]
```diff
@@ -82,7 +82,28 @@ function(_add_target_variant_c_compile_link_flags)
...
+          "${CMAKE_C_COMPILER}"
+          -Wno-unused-command-line-argument
+          -target-variant x86_64-apple-ios14.5-macabi -x c -c - -o /dev/null
+        INPUT_FILE
+          "${CMAKE_BINARY_DIR}/stdlib/empty"
+        OUTPUT_QUIET ERROR_QUIET
+        RESULT_VARIABLE
+          SUPPORTS_TARGET_VARIANT)
+
+      if(SUPPORTS_TARGET_VARIANT EQUAL 0)
```
Comment (compnerd) [inline stdlib/cmake/modules/AddSwiftStdlib.cmake:102]:
```suggestion
      if(NOT SUPPORTS_TARGET_VARIANT)
```

File: stdlib/cmake/modules/AddSwiftStdlib.cmake [hunk 3]
```diff
@@ -82,7 +82,28 @@ function(_add_target_variant_c_compile_link_flags)
       DEPLOYMENT_VERSION "${DEPLOYMENT_VERSION}")
     list(APPEND result "-target" "${target}")
     if(target_variant)
-      list(APPEND result "-target-variant" "${target_variant}")
+      # Check if the C compiler supports `-target-variant` flag
+      # TODO (etcwilde): This is a massive hack to deal with the fact that we
+      # are lying to cmake about what compiler is being used. Normally we could
+      # use `check_cxx_compiler_flag` here. Unfortunately, that uses a different
+      # compiler since we swap out the C compiler part way through the build.
```
Comment (compnerd) [inline stdlib/cmake/modules/AddSwiftStdlib.cmake:89]: I really wish that we would stop doing this ...

**Review by compnerd — APPROVED** [review]
This should be safe for Windows, but this is still pretty terrible.  The compiler swap occurs at stdlib build time right?

File: stdlib/cmake/modules/AddSwiftStdlib.cmake (hunk 3 above)
Comment (etcwilde) [reply stdlib/cmake/modules/AddSwiftStdlib.cmake:89]: I do too. I do too.

File: stdlib/cmake/modules/AddSwiftStdlib.cmake (hunk 1 above)
Comment (etcwilde) [reply stdlib/cmake/modules/AddSwiftStdlib.cmake:88]: I think it's `check_compiler_flag(<lang> ...)` now anyway. Will fix comment.

**Comment by etcwilde** [comment]
> The compiler swap occurs at stdlib build time right?

It happens in a couple of places. We swap the compilers for the stdlib, but it's also done in the bootstrap IIRC.

**Comment by etcwilde** [comment]
This is looking pretty good now. My local test with catalyst enabled is linking correctly but I'm still missing the catalyst swift modules. I'll keep digging.

**Comment by etcwilde** [comment]
@swift-ci please test

**Comment by etcwilde** [comment]
@swift-ci please macOS toolchain build

**Comment by etcwilde** [comment]
@swift-ci Please Build Toolchain macOS Platform

**Comment by etcwilde** [comment]
@swift-ci please test

**Comment by etcwilde** [comment]
Linux failure:
```
warning: 'swift-llbuild': skipping cache due to an error: Failed to clone repository https://github.com/apple/swift-llbuild.git:
    Cloning into bare repository '/home/build-user/build/buildbot_linux/.cache/org.swift.swiftpm/repositories/swift-llbuild-0f2964f9'...
    fatal: unable to access 'https://github.com/apple/swift-llbuild.git/': Failed to connect to github.com port 443: Connection timed out
```

Unrelated CI failure.

**Comment by etcwilde** [comment]
@swift-ci please test Linux

**Pull request merged**